proposals/p5366.md

```diff
@@ -0,0 +1,327 @@
+refined the rules for forward declarations, including the rules for `extern`
+declarations.
+
+[Leads issue #5251: `impl` declarations in a generic class context](https://github.com/carbon-language/carbon-lang/issues/5251)
+is (pending resolution) saying that an `impl` declaration in class scope must
+use `Self` in a deducible position.
+
+## Proposal
+
+An `impl` declaration is associated with the scope it is first declared in, and
+can only be redeclared in that scope, matching all other declarations. Consider
+how a function is redeclared outside the scope of a class in which it was
+originally defined:
+
+```
+class Z(T:! type) {
+  // Forward declaration of a function.
+  fn F();
+}
+
+// Definition of the function that was forward declared.
+fn Z(T:! type).F() { ... }
+```
+
+To redeclare an `impl` after the end of the scope it was declared in, that scope
+may be re-entered as part of the `impl` redeclaration, in the same way, except
+with parentheses around the name of the `impl`, as in:
+
+```carbon
+class X {
+  // Forward declaration that `X impls Y`:
+  impl as Y;
+}
+
+// Definition of the `impl` that `X impls Y`
+// that was forward declared in `X`:
+impl X.(as Y) { ... }
+```
+
+More generally, in a `class` scope
+
+```carbon
+class __X__ {
+  impl __Y__;
+}
+```
+
+is redeclared `impl __X__.(__Y__)` outside of that `class` scope. Here `__X__`
+is whatever sequence of tokens appears in that position in the `class`
+declaration, and `__Y__` is the sequence of tokens in the `impl` declaration.
+These declarations are matched syntactically, and anything in `__Y__` is
+interpreted as if it appeared in the scope of `__X__` like it was first
+declared.
+
+## Details
+
+Here are some examples of this in practice:
+
+```carbon
+class F {}
+
+class A {
+  impl as As(i32);
+  impl Self as As(bool);
+  impl A as As(f64);
+
+  impl F as As(A);
+
+  class G {}
+  impl G as As(A);
+}
+
+impl A.(as As(i32)) { ... }
+impl A.(Self as As(bool)) { ... }
+impl A.(A as As(f64)) { ... }
+impl A.(F as As(A)) { ... }
+impl A.(G as As(A)) { ... }
+```
+
+Parameterized classes:
+
+```carbon
+class B(T:! type) {
+  impl B(i32) as AddWith(A(T));
+}
+
+impl B(T:! type).(B(i32) as AddWith(A(T))) { ... }
+```
+
+Parameterized impl:
+
+```carbon
+class C {
+  impl forall [T:! type] as I(T);
+}
+
+impl C.(forall [T:! type] as I(T));
+```
+
+Putting the `forall` inside the parens both simplifies the syntactic match and
+means that any mentions of names in that clause are in scope. For example:
+
+```
+class D(T:! type) {
+  class E {}
+  impl forall [U:! J(E)] as I(U);
+}
+
+impl D(T:! type).(forall [U:! J(E)] as I(U));
+```
+
+Notice how the `E` in the constraint on `U` is found in the `D(T:! type)` scope.
+
+Nested classes:
+
+```
+class C1 {
+  class C2 {
+    class C3 {
+      impl as I;
+      class C4 {}
+    }
+  }
+}
+
+// Defining impl that was forward declared:
```

Comment:
I think this needs to be more specific?

```suggestion
// Defining impl that was forward declared within the `C3` definition:
```

Because we could declare at the file scope:

```carbon
impl C1.C2.C3.C4 as I;
```

And then define it later as you do below?